api/sendfild/service/service.go

```diff
@@ -0,0 +1,713 @@
+	FailureMsg    string             `bson:"failure_msg"`
+	CreatedAt     time.Time          `bson:"created_at"`
+	UpdatedAt     time.Time          `bson:"updated_at"`
+}
+
+func (t txn) latestMsgCid() (msgCid, error) {
+	if len(t.MessageCids) == 0 {
+		return msgCid{}, fmt.Errorf("no message cids found")
+	}
+	return t.MessageCids[len(t.MessageCids)-1], nil
+}
+
+type Service struct {
+	clientBuilder lotus.ClientBuilder
+	col           *mongo.Collection
+	server        *grpc.Server
+	waiting       map[primitive.ObjectID]chan waitResult
+	config        Config
+	ticker        *time.Ticker
+	waitingLck    sync.Mutex
+}
+
+type Config struct {
+	Listener           net.Listener
+	ClientBuilder      lotus.ClientBuilder
+	MongoUri           string
+	MongoDbName        string
+	MessageWaitTimeout time.Duration
+	MessageConfidence  uint64
+	RetryWaitFrequency time.Duration
+	Debug              bool
+}
+
+func New(ctx context.Context, config Config) (*Service, error) {
+	if config.Debug {
+		if err := util.SetLogLevels(map[string]logging.LogLevel{
+			"sendfil": logging.LevelDebug,
+		}); err != nil {
+			return nil, err
+		}
+	}
+
+	client, err := mongo.Connect(ctx, options.Client().ApplyURI(config.MongoUri))
+	if err != nil {
+		return nil, fmt.Errorf("connecting to mongo: %v", err)
+	}
+	db := client.Database(config.MongoDbName)
+	col := db.Collection(collectionName)
+	if _, err := col.Indexes().CreateMany(ctx, []mongo.IndexModel{
+		{
+			Keys: bson.D{primitive.E{Key: "from", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "to", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "amount_nano_fil", Value: 1}},
+		},
+		// MongoDB automatically creates a multikey index if any indexed field is an array;
+		// you do not need to explicitly specify the multikey type.
+		// https://docs.mongodb.com/manual/core/index-multikey/
+		{
+			Keys:    bson.D{primitive.E{Key: "message_cids.cid", Value: 1}},
+			Options: options.Index().SetUnique(true),
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "message_state", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "waiting", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "created_at", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "updated_at", Value: 1}},
+		},
+	}); err != nil {
+		return nil, fmt.Errorf("creating collection indexes: %v", err)
+	}
+
+	s := &Service{
+		clientBuilder: config.ClientBuilder,
+		col:           col,
+		waiting:       make(map[primitive.ObjectID]chan waitResult),
+		config:        config,
+		ticker:        time.NewTicker(config.RetryWaitFrequency),
+	}
+
+	s.server = grpc.NewServer()
+	go func() {
+		pb.RegisterSendFilServiceServer(s.server, s)
+		if err := s.server.Serve(config.Listener); err != nil && !errors.Is(err, grpc.ErrServerStopped) {
+			log.Errorf("serve error: %v", err)
+		}
+	}()
+
+	if err := s.waitAllQualifying(ctx, true); err != nil {
+		return nil, fmt.Errorf("calling waitAllQualifying: %v", err)
+	}
+
+	s.bindTicker(ctx)
```

Comment:
We bootstrap our message waiting on init and then bind the ticker so we retry message waiting periodically.